### PR TITLE
Fix register device command

### PR DESCRIPTION
**Pull request by esdrasjnr**
What does this implement/fix? Explain your changes
---------------------------------------------------
This patch fixed the actual wrong register device flow and the json field of the list devices response message

Where has this been tested?
---------------------------
**Operating System/Platform:** Ubuntu 18.04.4

**Go Version:** 1.13.6

### CODE REVIEW

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- pkg/thing/interactors/unregister_thing.go  7
- pkg/thing/interactors/register_thing.go  1
- pkg/thing/interactors/unregister_thing_test.go  3
- pkg/thing/delivery/http/thing_proxy.go  1
- pkg/thing/entities/errors.go  1
- pkg/thing/handler/amqp/msg_handler.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5153954&bid=16927011)

File: pkg/thing/interactors/register_thing.go
Comment (ramonhpr) [inline pkg/thing/interactors/register_thing.go]:
```suggestion
			i.logger.Error(err)
```

File: pkg/thing/interactors/register_thing.go
Comment (ramonhpr) [inline pkg/thing/interactors/register_thing.go]:
```suggestion
			i.logger.Error(err)
```

File: pkg/thing/interactors/register_thing.go
Comment (ramonhpr) [inline pkg/thing/interactors/register_thing.go]: You can remove this call since it is already outside the if statement

File: pkg/thing/interactors/register_thing.go
Comment (ramonhpr) [reply pkg/thing/interactors/register_thing.go]: This log won't be showed if the log level is INFO or ERROR just if it is DEBUG

**Review by ramonhpr — DISMISSED** [review]

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- pkg/thing/interactors/register_thing.go  1
- pkg/thing/delivery/http/thing_proxy.go  1
- pkg/thing/entities/errors.go  1
- pkg/thing/handler/amqp/msg_handler.go  2
- pkg/thing/interactors/unregister_thing.go  7
- pkg/thing/interactors/unregister_thing_test.go  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5153954&bid=16927011)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- pkg/thing/handler/amqp/msg_handler.go  2
- pkg/thing/delivery/http/thing_proxy.go  1
- pkg/thing/interactors/unregister_thing.go  7
- pkg/thing/interactors/unregister_thing_test.go  3
- pkg/thing/interactors/register_thing.go  1
- pkg/thing/entities/errors.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5153954&bid=16927011)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- pkg/thing/entities/errors.go  1
- pkg/thing/handler/amqp/msg_handler.go  2
- pkg/thing/interactors/unregister_thing_test.go  3
- pkg/thing/interactors/register_thing.go  1
- pkg/thing/delivery/http/thing_proxy.go  1
- pkg/thing/interactors/unregister_thing.go  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5153954&bid=16927011)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/delivery/http/thing_proxy.go  1
- pkg/thing/entities/errors.go  1
- pkg/thing/interactors/register_thing.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5153954&bid=16927011)

File: pkg/network/msg.go
Comment (lcbm) [inline pkg/network/msg.go]: Is there a specific reason for renaming `things` in favor of `devices`? Unless there is, considering that throughout the project we have been referring to this `object` as `thing` (instead of `device`), it'd be nice to keep it standardized (as it was before). :grimacing:

File: pkg/thing/delivery/amqp/client.go
Comment (lcbm) [inline pkg/thing/delivery/amqp/client.go]: Same as [here](https://github.com/CESARBR/knot-babeltower/pull/46#discussion_r390567459) :smile:

File: pkg/thing/interactors/register_thing.go
Comment (lcbm) [inline pkg/thing/interactors/register_thing.go]: We should avoid returning statements such as this one. Modifying [Dave Chaney's example](https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully) to our context :point_down: 

> If `i.connectorPublisher.SendRegisterDevice(id, name)` returns an error, then `Register` will return the error to its caller, who will probably do the same, and so on. At the top of the program the main body of the program will print the error to the screen or a log file, and all that will be printed is: "No such file or directory," for example. In other words, we cannot tell where the original error came from.

Here's an example :smile: 

```suggestion
	err = i.connectorPublisher.SendRegisterDevice(id, name)
	if err != nil {
		i.logger.Error(err)
		return err
	}

	return nil
}
```

However, if we must return `i.connectorPublisher.SendRegisterDevice(id, name)`, it'd be nice to formalize and standardize when and why to do this (such as in [here](https://github.com/CESARBR/knot-babeltower/blob/master/pkg/thing/delivery/http/thing_proxy.go#L201)).

File: pkg/thing/interactors/register_thing.go
Comment (lcbm) [inline pkg/thing/interactors/register_thing.go]: What do you think about rearranging the order of these operations, so that we only reply to the `client` once the `connector` publish operation is executed and returns an error? 

Here's a suggestion :tada: 

```go
	err = i.connectorPublisher.SendRegisterDevice(id, name)
	if err != nil {
		// handle error here
		i.logger.Error(err)
	}

	err = i.reply(id, token, err)
	if err != nil {
		i.logger.Error(err)
		return err
	}

	return nil
```

Also, note the code doesn't return `i.reply(id, token, err)` -- instead, it "handles" the error and returns `nil`. I left a comment [here](https://github.com/CESARBR/knot-babeltower/pull/46#discussion_r390609437) to explain why :grimacing:

File: pkg/thing/interactors/register_thing.go
Comment (lcbm) [reply pkg/thing/interactors/register_thing.go]: I didn't quite catch why, could you explain? :eyes: 

From my understanding, if there is only one `i.reply(id, token, errorVariable)`, `Register()` wouldn't be able to notify the client about either: 

1. the registration error in case `i.thingProxy.Create(id, name, authorization)` 
2. the error in case `i.connectorPublisher.SendRegisterDevice(id, name)`

**Review by lcbm — CHANGES_REQUESTED** [review]
Nice PR! Most comments I left are actually doubts, except for one :stuck_out_tongue:

File: pkg/thing/interactors/register_thing.go
Comment (lcbm) [reply pkg/thing/interactors/register_thing.go]: I didn't quite catch why, could you explain? :eyes: 

From my understanding, if there is only one `i.reply(id, token, errorVariable)`, `Register()` wouldn't be able to notify the client about either: 

1. the registration error in case `i.thingProxy.Create(id, name, authorization)` 
2. the error in case `i.connectorPublisher.SendRegisterDevice(id, name)`

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/entities/errors.go  1
- pkg/thing/interactors/register_thing.go  1
- pkg/thing/delivery/http/thing_proxy.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5153954&bid=16927011)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/entities/errors.go  1
- pkg/thing/interactors/register_thing.go  1
- pkg/thing/delivery/http/thing_proxy.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5153954&bid=16927011)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/interactors/register_thing.go  1
- pkg/thing/entities/errors.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5153954&bid=16927011)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/interactors/register_thing.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5153954&bid=16927011)

File: pkg/network/msg.go
Comment (netoax) [reply pkg/network/msg.go]: +1
If you just want to change the JSON property name, use the tags for that. That's why they were created for.

File: pkg/thing/interactors/register_thing.go
Comment (netoax) [reply pkg/thing/interactors/register_thing.go]: He was talking about the same method being called [here](https://github.com/CESARBR/knot-babeltower/pull/46/commits/de95a6815fa490dda27198ce1028eb0eeeb8a8cc#diff-60df0fe9cfe391c1945e90caf147ab91R102).

File: pkg/thing/interactors/register_thing.go
Comment (netoax) [reply pkg/thing/interactors/register_thing.go]: +1
You're right. However, in this case, we will get the same result.

The point about handling errors as Dave Cheney suggested is that we have to append contextual information to the error before return it (@esdrasjnr I strongly suggest you read the entire session of the article @lcbm mentioned).

In a lot of cases, we don't append any contextual information to the errors. Let's change this incrementally.

File: pkg/thing/interactors/register_thing.go
Comment (netoax) [reply pkg/thing/interactors/register_thing.go]: +1
Much better.

**Review by netoax — CHANGES_REQUESTED** [review]

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/interactors/register_thing.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5153954&bid=16927011)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/interactors/register_thing.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5153954&bid=16927011)

File: pkg/network/msg.go
Comment (esdrasjnr) [reply pkg/network/msg.go]: I verify the [docs](https://github.com/CESARBR/knot-fog-connector/blob/master/docs/api/amqp.md#devices-list) and notice that the expected JSON property was `devices` instead of `things`. So I changed the struct variable name either to be concised with the nomenclature. What do you thought about it?

File: pkg/thing/delivery/amqp/client.go
Comment (esdrasjnr) [reply pkg/thing/delivery/amqp/client.go]: This was necessary to reference the new renamed structure :confused:

File: pkg/thing/interactors/register_thing.go
Comment (esdrasjnr) [reply pkg/thing/interactors/register_thing.go]: @lcbm @joaoaneto Excellent article, thanks! In order to add some context to errors, I made some changes in code. Could you please check if it makes sense? :smiley:

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/interactors/register_thing.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5153954&bid=16927011)

File: pkg/thing/interactors/register_thing.go
Comment (esdrasjnr) [reply pkg/thing/interactors/register_thing.go]: I didn't do this because this change break the tests. However, I agree that it's better to understand and also will be according with `unregister` interactor.

File: pkg/thing/interactors/register_thing.go
Comment (esdrasjnr) [reply pkg/thing/interactors/register_thing.go]: @ramonhpr You are right! I replace the method call to right after the `Create` method call. Could you check if it's ok? :smile:

File: pkg/thing/interactors/register_thing.go
Comment (esdrasjnr) [reply pkg/thing/interactors/register_thing.go]: In order to add some context to the errors, I've made some changes, including removing the debug logging in this interactor function. Done!

File: pkg/thing/interactors/register_thing.go
Comment (esdrasjnr) [reply pkg/thing/interactors/register_thing.go]: Done!

File: pkg/thing/interactors/register_thing.go
Comment (esdrasjnr) [reply pkg/thing/interactors/register_thing.go]: I've made some changes in patch, could you review it please? :+1:

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/interactors/register_thing.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5153954&bid=16927011)

File: pkg/network/msg.go
Comment (netoax) [reply pkg/network/msg.go]: Personally, I prefer to use the term "thing" to talk about smart devices. We often use "devices" because that's the way everyone refers to since starting to work on the KNoT project.

In terms of code, the tags we attach to the structure exists exactly to be used when we have different representations between our internal naming and the property name we want to export in the JSON.

File: pkg/thing/interactors/register_thing.go
Comment (netoax) [reply pkg/thing/interactors/register_thing.go]: @esdrasjnr did you push this change?

File: pkg/thing/interactors/register_thing.go
Comment (netoax) [inline pkg/thing/interactors/register_thing.go]: Please, follow the DRY (Don't Repeat Yourself) rule. Notice what you're doing here is quite similar to lines +85 to +88. My suggestions:

- Let's take advantage of the fact the method `reply` is already expecting an error as one of its attributes. i.e. we can append the sending error information there instead of doing this in the `Register` method. For example:

```go
	err := i.verifyThingID(id)
	if err != nil {
		return i.reply(id, "", fmt.Errorf("invalid id: %w", err))
	}
```

We add context to the ID validation error and pass it to the `reply`, which adds the context information about the sending error. After that, we just need to return directly.

- Using the same structure (when creating the thing and sending the response despite success or failure):

```go
	token, err := i.thingProxy.Create(id, name, authorization)
	err = i.reply(id, token, err) // reply() will add the necessary context information and we try to send the response (success or failure).
	if err != nil {
                /* The response can't be sent and as the error already came with the sending error context information, 
                 we just need to add a bit of information about the whole procedure and return it. */
		return fmt.Errorf("error registering thing: %w", err)
	}
```

Comment (netoax) [reply pkg/thing/interactors/register_thing.go]: Also, you have a good opportunity to refactor the `reply` method. Is "reply" a good name? Couldn't we move the part of it which is responsible for getting a pointer string from a received error to another function, something like `i.getNullableErrString(err)`?.

Comment (netoax) [reply pkg/thing/interactors/register_thing.go]: Couldn't we change the `verifyThingID` to get rid of `ErrorIDLenght` and `ErrorIDInvalid` ? `errors.New()` is utterly  fine here.

Comment (netoax) [reply pkg/thing/interactors/register_thing.go]: Couldn't we change the order of the functions? The majority of the code we're writing is placing the most important function at the beginning of the file. i.e. from the most important to the helpers, in a way that the code can be read as a text.

**Review by netoax — CHANGES_REQUESTED** [review]

File: pkg/thing/interactors/register_thing.go
Comment (esdrasjnr) [reply pkg/thing/interactors/register_thing.go]: @joaoaneto Yes! Note that I only replace the `reply` function call. This was necessary to maintain the desired flow. However, the changes you suggested will change it soon :smiley:

File: pkg/network/msg.go
Comment (lcbm) [reply pkg/network/msg.go]: I agree with @joaoaneto :smile:

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/interactors/register_thing.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5153954&bid=16927011)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/interactors/register_thing.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5153954&bid=16927011)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- pkg/thing/delivery/amqp/client.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5153954&bid=16927011)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- pkg/thing/delivery/amqp/client.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5153954&bid=16927011)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- pkg/thing/delivery/amqp/client.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5153954&bid=16927011)

File: pkg/thing/interactors/register_thing.go
Comment (esdrasjnr) [reply pkg/thing/interactors/register_thing.go]: @joaoaneto Excellent points! I've made some adjustments, waiting for review! :smile:

File: pkg/thing/interactors/register_thing.go
Comment (netoax) [inline pkg/thing/interactors/register_thing.go]: "it was not possible to create a thing, so returns without sending a response to the connector".

File: pkg/thing/interactors/register_thing.go
Comment (netoax) [inline pkg/thing/interactors/register_thing.go:33]: Isn't this error handled already on the `sendResponse()` method?

**Review by netoax — CHANGES_REQUESTED** [review]

File: pkg/thing/interactors/register_thing.go
Comment (netoax) [reply pkg/thing/interactors/register_thing.go]: Nice @esdrasjnr, the code improved a lot. Thanks!

File: pkg/thing/interactors/register_thing.go
Comment (esdrasjnr) [reply pkg/thing/interactors/register_thing.go:33]: @joaoaneto Yes, it is handled in sendResponse, but it still needs to be returned in the case `sendErr != nil`. However, in line 35, the `sendError` should be returned instead of `nil`.

Changes done!

File: pkg/thing/interactors/register_thing.go
Comment (esdrasjnr) [reply pkg/thing/interactors/register_thing.go]: Done!

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- pkg/thing/delivery/amqp/client.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5153954&bid=16927011)

File: pkg/thing/interactors/register_thing.go
Comment (lcbm) [inline pkg/thing/interactors/register_thing.go]: Nice, also had a feeling about these errors :joy: 

I also saw [this](https://github.com/docker/distribution/blob/master/reference/reference.go#L40) approach and it seems reasonable:

1. Erros are declared on the top of the file (better readability, lookup and scaling)
2. Erros are referenced in a more concise way (better readability and easier to manage when refactoring) 

What do you think about doing something like this? I guess its a middle ground between as it was before and this new approach. Also, the approach referenced seems to have a few advantages worth considering.

File: pkg/thing/interactors/register_thing.go
Comment (lcbm) [inline pkg/thing/interactors/register_thing.go]: I was actually wondering if this is the most idiomatic way to get values from repetitive clauses and apparently it is. Just leaving this as a +1 here :smile: 
> golang: https://golang.org/doc/faq#Does_Go_have_a_ternary_form
> stackoverflow: https://stackoverflow.com/questions/19979178/what-is-the-idiomatic-go-equivalent-of-cs-ternary-operator

I'd, personally consider actually using an if-else statement at the top of the function calling `getErrMessagePtr()` (given its only used by one function ATM), instead of declaring a function -- but this is just personal preference.

**Review by lcbm — COMMENTED** [review]
Awesome :smile: 

I made two small comments, let me know when you take a look at it :+1:

File: pkg/thing/interactors/register_thing.go
Comment (esdrasjnr) [reply pkg/thing/interactors/register_thing.go]: @lcbm Agree with you! I think it's better because in test files we expect these errors to occur in use cases

File: pkg/thing/interactors/register_thing.go
Comment (netoax) [reply pkg/thing/interactors/register_thing.go]: In terms of clean code, the better approach is to break that function. It was so hard to understand and it did more than one thing (despite being simple or not). At least for me, with `getErrMessagePtr()` the function calling it became clearer.

File: pkg/thing/interactors/register_thing.go
Comment (netoax) [reply pkg/thing/interactors/register_thing.go]: +1 on this approach, @lcbm.

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- pkg/thing/delivery/amqp/client.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5153954&bid=16927011)

File: pkg/thing/delivery/amqp/client.go
Comment (lcbm) [inline pkg/thing/delivery/amqp/client.go]: Ooopss :grimacing: 
I think this is a leftover from `git rebase` :joy: 

```suggestion
func (mp *msgClientPublisher) SendRegisteredDevice(thingID, token string, errMsg *string) error {
```

**Review by lcbm — CHANGES_REQUESTED** [review]
Awesome PR! :smile: 
Theres some leftover from `git rebase`, tho! :joy: 
Let me know once its fixed, so that I may approve the PR :smile:

File: pkg/thing/interactors/register_thing.go
Comment (esdrasjnr) [reply pkg/thing/interactors/register_thing.go]: @lcbm @joaoaneto Done!

File: pkg/thing/delivery/amqp/client.go
Comment (esdrasjnr) [reply pkg/thing/delivery/amqp/client.go]: @lcbm I'm also forgot to rename the msg struct to `DeviceRegisteredResponse`. Now, all done! :smile:

**Review by lcbm — APPROVED** [review]
LGTM :confetti_ball:

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- pkg/thing/delivery/amqp/client.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5153954&bid=16927011)

**Review by netoax — APPROVED** [review]
LGTM, thanks!

**Comment by netoax** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- pkg/thing/delivery/amqp/client.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5153954&bid=16927011)

**Pull request merged**
